sonic_platform_base/card_base.py

```diff
@@ -0,0 +1,80 @@
+#
+# card_base.py
+#
+# Abstract base class for implementing a platform-specific class with which
+# to interact with a card in SONiC chassis
+#
+
+class CardBase(object):
```

Comment:
The existing `ModuleBase` class was designed as a placeholder base class for supporting removable cards (i.e., "modules") in a modular chassis. Can we not extend or inherit from that class? This class appears to make `ModuleBase` obsolete.

Reply:
It seemed like ModuleBase was designed for a non-disaggregated chassis i.e. if line-cards were not having a CPU and their full representation needed to be on the control-card. Intention is not to make ModuleBase obsolete. The CardBase is complementary.

As suggested, we can make it a derived class retaining much of the definition. Does class ModuleCardBase(ModuleBase) work? Other suggestions are welcome.

Reply:
Do the APIs in ModuleBase also apply to what you consider a "card" here? The concept of a "Module" was to represent any pluggable card. If you feel there is a difference, can you please explain further what the differentiation is?

Reply:
The APIs of ModuleBase did not seem to apply since for Modular Chassis the ChassisBase covers what ModuleBase does. 
In CardBase we want to represent just the state, type and instance(slot number) of each component like line-card, fabric-card individually. This would be similar to FanBase and PsuBase for example. 

We did deliberate on whether to re-use. If you think we should roll the below APIs in ModuleBase and re-use whatever DeviceBase will provide like status() etc, we can.

Reply:
At the very least, it should inherit from `DeviceBase`. ModuleBase was created to assume that a card could contain an EEPROM, fans, thermals, components (e.g., BIOS, FPGA, CPLD, etc.) and potentially a dedicated power supply. Since you're more familiar with the types of cards in use, is this a correct assumption?

Reply:
Card does have eeprom, thermals, components. I think the confusion arises with the disaggregated chassis where a sonic instance runs on the card and all of these get covered by chassis.py. So, seemed redundant unless we want a full line-card (or other cards) representation on Control-Card.

Is it okay if we add the missing APIs from CardBase to ModuleBase and make progress?

Reply:
> Is it okay if we add the missing APIs from CardBase to ModuleBase and make progress?

Yes, please go ahead.

Reply:
With plan to leverage ModuleBase, can you please update/clarify on the plan to use following existing APIs of chassisBase.py instead of the new ones suggested/listed here?

Use following : get_num_modules() ; get_all _modules() ; get_module() etc.
instead of adding following new helper APIs in chassis_base.py
    get_num_cards(),
    get_card()
    get_all_cards()
    get_controlcard_instance()
    get_my_instance()

Reply:
Yes, we will re-use what the ModuleBase.py provides. All APIs introduced in ChassisBase.py will remain as is.